core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2601,11 +2601,20 @@ object ZIO {
 
   /**
    * Merges an `Iterable[IO]` to a single IO, working in parallel.
+   *
+   * It's possible to execute side effects inside `f`, as `f` is guaranteed to
+   * be executed at most once for each `in` element.
    */
   def mergeAllPar[R, E, A, B](
     in: Iterable[ZIO[R, E, A]]
   )(zero: B)(f: (B, A) => B): ZIO[R, E, B] =
-    in.foldLeft[ZIO[R, E, B]](succeed[B](zero))((acc, a) => acc.zipPar(a).map(f.tupled)).refailWithTrace
+    RefM.make(zero) >>= { acc =>
```

Comment:
Let’s rewrite this with Ref instead.

Reply:
Then `f` will be retried on high contention, and it will be unsafe to pass side-effecting functions. Are we sure it's OK to change this predicate? Old `mergeAllPar` version doesn't retry `f`.

Reply:
Perfectly fine. The ZIO API makes no accommodations for side effecting functions when pure functions are expected.

Reply:
e.g., this test doesn't pass if rewrite with Ref: https://github.com/zio/zio/pull/2614/files#diff-ce160572c9127744730d5a14dc428162R746-R755

Reply:
Right. You should rewrite it without side effects in f.